packages/arb-validator/checkpointing/indexedCheckpointer.go

```diff
@@ -246,49 +150,36 @@ func (cp *IndexedCheckpointer) AsyncSaveCheckpoint(
 	}
 }
 
-func (_ *IndexedCheckpointer) makeContentsKey(id *common.BlockId) []byte {
-	bidBuf := id.MarshalToBuf()
-	bytesBuf, err := proto.Marshal(bidBuf)
-	if err != nil {
-		log.Fatal(err)
-	}
-	return append([]byte{2}, bytesBuf...)
-}
-
-func (cp *IndexedCheckpointer) RestoreLatestState(ctx context.Context, clnt arbbridge.ArbClient, unmarshalFunc func([]byte, RestoreContext) error) error {
+func (cp *IndexedCheckpointer) RestoreLatestState(ctx context.Context, clnt arbbridge.ChainTimeGetter, unmarshalFunc func([]byte, RestoreContext) error) error {
 	cp.Lock()
 	defer cp.Unlock()
 
-	heightBounds, err := cp.getHeightBounds()
-	if err != nil {
-		return err
+	if cp.db.IsBlockStoreEmpty() {
+		return errNoCheckpoint
 	}
-	if heightBounds == nil {
-		return errors.New("Cannot restore because no checkpoint exists")
-	}
-	for height := heightBounds.hi; height.Cmp(heightBounds.lo) >= 0; height = common.NewTimeBlocks(new(big.Int).Sub(height.AsInt(), big.NewInt(1))) {
-		ids, err := cp.getIdsAtHeight(height)
+
+	startHeight := cp.db.MaxBlockStoreHeight()
+	lowestHeight := cp.db.MinBlockStoreHeight()
+
+	for height := startHeight; height.Cmp(lowestHeight) >= 0; height = common.NewTimeBlocks(new(big.Int).Sub(height.AsInt(), big.NewInt(1))) {
+		onchainId, err := clnt.BlockIdForHeight(ctx, height)
 		if err != nil {
 			return err
 		}
-		onchainId, err := clnt.BlockIdForHeight(ctx, height)
+		blockData, err := cp.db.GetBlock(onchainId)
 		if err != nil {
-			return err
+			// If no record was found, try the next block
+			continue
 		}
-		for _, id := range ids {
-			if id.Equals(onchainId) {
-				key := cp.makeContentsKey(id)
-				val := cp.db.GetData(key)
-				ckpWithMan := &CheckpointWithManifest{}
-				if err := proto.Unmarshal(val, ckpWithMan); err != nil {
-					return err
-				}
-				return unmarshalFunc(ckpWithMan.Contents, cp.newRestoreContextLocked())
-			}
+		ckpWithMan := &CheckpointWithManifest{}
+		if err := proto.Unmarshal(blockData, ckpWithMan); err != nil {
+			// If something went wrong, try the next block
+			continue
 		}
+		return unmarshalFunc(ckpWithMan.Contents, cp.newRestoreContextLocked())
+
 	}
-	log.Fatal("Called RestoreLatestState on checkpointer that has no stored checkpoints")
-	return nil // can't reach this but need to make the compiler happy
+	return errNoMatchingCheckpoint
 }
 
 func (cp *IndexedCheckpointer) writeDaemon() {
```

Comment:
The lock handling in writeDaemon is inefficient.  (Not a problem with this PR; was pre-existing.) I'll make a code commit to fix this.  

The problem is that writeDaemon holds the lock while it is writing out the old checkpoint. This blocks calls to AsyncSaveCheckpoint, which blocks the rollup manager's processing of new on-chain blocks.

Reply:
I've massively reduced the amount of locking. You should take a look and see if it seems reasonable